### PR TITLE
Issue 243 face xy topology

### LINKED ISSUE #243: Add face/xy topology primitives for HEALPix NESTED cells
# Add face/xy topology primitives for HEALPix NESTED cells

## Motivation

`healpix-analyse` is prototyping face-native spatial filtering for Sentinel-2 processing, including a separable Gaussian filter.

Efficient face-native processing requires two low-level HEALPix NESTED operations:

1. conversion between a NESTED cell ID and its base-face-local `(x, y)` coordinates;
2. deterministic remapping of local coordinates when crossing a HEALPix base-face boundary.

These are pure HEALPix topology operations. They require no longitude/latitude conversion, ellipsoid, projection, or geodesic computation.

`healpix_geo.nested` currently exposes geographic conversion and coverage functions, but no public face-local topology primitives.

Since `healpix-geo` already builds on `cds-healpix-rust`, these operations should reuse the existing NESTED/Z-order and neighbour-topology machinery there wherever possible, rather than introducing an independent HEALPix implementation in Python or duplicating Morton-code logic.

The goal of this issue is therefore to expose a small, fast, well-tested topology API from `healpix-geo` that can also be reused by other face-native algorithms.

---

## 1. NESTED cell ↔ `(face, x, y)`

Add public vectorized functions equivalent in purpose to the established HEALPix `pix2xyf` / `xyf2pix` operations:

```python
healpix_geo.nested.pix2xyf(cell_ids, level)
    -> (face, x, y)

healpix_geo.nested.xyf2pix(face, x, y, level)
    -> cell_ids
```

For HEALPix level `level`:

```text
nside = 2**level
```

and each of the 12 base faces contains an `nside × nside` integer grid.

The NESTED cell index is conceptually composed of:

```text
base-face id + Morton/Z-order encoding of (x, y)
```

so these functions should be implemented using integer operations only.

### Implementation requirement

Do **not** introduce a second independent Morton/Z-order implementation if the required primitives can be obtained from `cds-healpix-rust`.

Prefer using or exposing the existing `cds-healpix-rust` NESTED/Z-order implementation and adding only the minimal `healpix-geo` wrapper/API required.

### Expected ranges

```text
face: 0 .. 11
x:    0 .. nside - 1
y:    0 .. nside - 1
```

The API should support NumPy arrays / batched input consistently with the existing `healpix_geo.nested` API.

No `ellipsoid` parameter is required.

---

## 2. Base-face neighbour transforms

HEALPix base faces do not all share the same local `(x, y)` orientation.

When an operation crosses a base-face boundary, the target face may therefore require a combination of:

- swapping `x` and `y`;
- reversing `x`;
- reversing `y`.

This transformation depends on both:

```text
source face
+
boundary-crossing direction
```

and should not be inferred only from whether the face belongs to the north-polar, equatorial, or south-polar group.

Expose a topology primitive such as:

```python
healpix_geo.nested.face_neighbour_transform(face, direction)
    -> transform
```

where `direction` represents one of the eight HEALPix neighbour directions.

A returned transform should contain enough information to remap coordinates without geographic calculations, for example conceptually:

```text
target_face
swap_xy
flip_x
flip_y
```

The exact public representation may follow the conventions that fit best with the existing Rust/Python API.

The important requirement is that callers must not need to reimplement HEALPix face adjacency or orientation rules themselves.

### Source of truth

The face adjacency and orientation rules should be derived from the existing HEALPix topology implementation, preferably reusing the corresponding machinery already present in `cds-healpix-rust`.

They should **not** be reconstructed from longitude/latitude samples or independently guessed from diagrams.

The official HEALPix neighbour algorithm should be treated as the reference topology.

---

## 3. Optional coordinate-remapping helper

If it provides a cleaner and safer interface, consider also exposing an internal or public helper that directly maps an offset coordinate across base-face boundaries, e.g.:

```python
healpix_geo.nested.transform_xy_across_face(
    face,
    x,
    y,
    dx,
    dy,
    level,
)
    -> (target_face, target_x, target_y)
```

This is not necessarily required for the first implementation if the lower-level transform API is sufficient.

However, such a helper could prevent downstream packages such as `healpix-analyse` from having to reproduce multi-edge and corner traversal logic.

---

## 4. Why this belongs in `healpix-geo`

The intended use case is face-native filtering such as:

```text
HEALPix face
    ↓
local 2-D operation
    ↓
boundary padding
    ↓
topological remapping into neighbouring face(s)
```

The face-boundary operation is HEALPix geometry/topology infrastructure, not Gaussian-filter-specific logic.

Putting it in `healpix-geo` therefore:

- avoids duplication across downstream packages;
- keeps HEALPix topology out of `healpix-analyse`;
- allows the implementation to use the Rust HEALPix backend directly;
- provides a reusable primitive for convolution, morphology, local neighbourhood operations, tiling, and other face-native algorithms.

---

### 5.1 Round-trip tests

For all 12 base faces and multiple levels:

```python
pix == xyf2pix(*pix2xyf(pix, level), level)
```

Test at least:

```text
level 0
level 1
level 2
several intermediate levels
a high supported level
```

Level 0 must be covered explicitly.

At level 0:

```text
nside = 1
x = 0
y = 0
cell id == base-face id
```

---

### 5.2 Exhaustive small-level tests

For small levels where exhaustive testing is inexpensive, test every pixel on every face.

Verify:

```text
cell -> face/x/y -> cell
```

and

```text
face/x/y -> cell -> face/x/y
```

---

### 5.3 Independent cross-check

Cross-check the conversion API against an independent HEALPix implementation such as `healpy`:

```python
healpy.pix2xyf(..., nest=True)
healpy.xyf2pix(..., nest=True)
```

for representative and randomly generated NESTED cells.

This should be a test/reference dependency only; the production implementation must not depend on `healpy`.

---

### 5.4 Face-edge transforms

Test every base face and every valid crossing direction.

Coverage must include:

- north-polar ↔ north-polar boundaries;
- north-polar ↔ equatorial boundaries;
- equatorial ↔ equatorial boundaries;
- equatorial ↔ south-polar boundaries;
- south-polar ↔ south-polar boundaries.

Verify both:

```text
target face
```

and:

```text
target coordinate orientation
```

---

### 5.5 Corner cases

Explicitly test base-face corners where more than two faces meet.

Corner behaviour must agree with the canonical HEALPix neighbour topology and must not depend on an arbitrary traversal order.

This is especially important for later convolution/padding use cases.

---

### 5.6 Neighbour equivalence

Where possible, verify that:

```text
face/xy boundary transform
        ↓
xyf2pix
```

produces the same neighbouring NESTED cell IDs as the established HEALPix neighbour implementation.

This provides an end-to-end validation of the orientation table.

---

### 5.7 Vectorized behaviour

Test:

- scalar input;
- 1-D arrays;
- multidimensional arrays;
- supported broadcasting behaviour;
- dtype and output-shape consistency.

---

### 5.8 Invalid inputs

Test validation for at least:

```text
face outside 0..11
x >= nside
y >= nside
invalid level
cell id outside the valid range for the requested level
```

---

## 6. Implementation guidance

Recommended implementation hierarchy:

```text
HEALPix specification
        ↓
official HEALPix neighbour topology
        ↓
cds-healpix-rust implementation
        ↓
healpix-geo Rust API
        ↓
Python bindings
        ↓
healpix-analyse
```

In particular:

- reuse the `cds-healpix-rust` NESTED/Z-order implementation where possible;
- reuse its neighbour/topology primitives where possible;
- avoid longitude/latitude round-trips;
- avoid floating-point calculations for these APIs;
- avoid duplicating HEALPix face-orientation logic in Python.

---

## 7. Acceptance criteria

- [ ] `healpix_geo.nested.pix2xyf(...)` implemented.
- [ ] `healpix_geo.nested.xyf2pix(...)` implemented.
- [ ] Conversion uses integer NESTED/Z-order operations only.
- [ ] Existing `cds-healpix-rust` functionality is reused where applicable.
- [ ] Base-face neighbour/orientation transform API implemented.
- [ ] All 12 faces and all relevant crossing directions covered.
- [ ] Face corners explicitly tested.
- [ ] Round-trip tests pass at level 0 and multiple higher levels.
- [ ] Results cross-checked against an independent HEALPix implementation.
- [ ] Vectorized NumPy API documented and tested.
- [ ] No geographic/geodesic conversion is used internally for these operations.

---

## References

- Górski, K. M. et al. (2005), *HEALPix: A Framework for High-Resolution Discretization and Fast Analysis of Data Distributed on the Sphere*, ApJ 622, 759–771.
- Calabretta, M. R. (2004), *Mapping on the HEALPix grid*.
- HEALPix official `neighbours_nest` documentation / implementation.
- `cds-healpix-rust` NESTED and Z-order implementation.
- `healpy.pix2xyf` / `healpy.xyf2pix` as an independent validation reference.

## Related

- `GRID4EARTH/healpix-analyse` — face-native Gaussian / spatial filtering work.
- `GRID4EARTH/healpix-geo#242` — related effort to move repeated low-level HEALPix work into `healpix-geo`.

### PR DESCRIPTION
**Pull request by tinaok**
see PR message at https://github.com/GRID4EARTH/healpix-geo/pull/244

close #243

### CODE REVIEW

File: healpix-geo-python/python/healpix_geo/tests/test_topology.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    (None, 1, 6, 9, None, 8, 4, 0),
+    (None, 2, 7, 10, None, 9, 5, 1),
+    (None, 3, 4, 11, None, 10, 6, 2),
+    (0, 5, None, 9, 10, 11, None, 4),
+    (1, 6, None, 10, 11, 8, None, 5),
+    (2, 7, None, 11, 8, 9, None, 6),
+    (3, 4, None, 8, 9, 10, None, 7),
+)
+
+
```
Comment (keewis) [inline healpix-geo-python/python/healpix_geo/tests/test_topology.py:22]: please put the test cases into grouped classes if there's more than two test functions per tested function. For this one, it would be (but see also the other comments on potentially renaming the function):

```suggestion

class TestFaceNeighbourTransform:
```

that also means we don't need the `_face_neighbour_transform_` part in the name of the test functions.

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+    Some(FaceTransform {
+        target_face,
+        swap_xy,
+        flip_x,
+        flip_y,
+    })
+}
+
+/// Split NESTED cell indexes into base-face-local coordinates.
```
Comment (keewis) [inline rust/healpix-geo-core/src/vectorized/nested/topology.rs]: no need to talk about nested, this entire crate indicates the context:
```suggestion
/// Split cell indexes into base-face-local coordinates.
```

File: python/healpix-geo/python/healpix_geo/nested.py [hunk 3]
```diff
@@ -17,10 +17,171 @@
...
+    swap_xy: bool
+    flip_x: bool
+    flip_y: bool
+
+
 def create_empty(depth):
     return RangeMOCIndex.empty(depth)
 
 
+def face_neighbour_transform(face, direction):
```
Comment (keewis) [inline python/healpix-geo/python/healpix_geo/nested.py]: what is the purpose of this function? For comparison, none of the existing healpix implementations appear to expose this (as far as I can tell, I didn't search too deeply).

I'd like to understand the context, because that may help me judge whether it would be better to consider that task or keep this one (I didn't check the earlier PR yet, that may contain the information).

However, if we do keep it I would argue that `FaceTransform` should actually be a callable that uses the array API to transform the passed object.

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    } else {
+        (dx.0 < 0, dy.1 < 0)
+    };
+
+    Some(FaceTransform {
+        target_face,
+        swap_xy,
+        flip_x,
+        flip_y,
+    })
```
Comment (keewis) [inline rust/healpix-geo-core/src/vectorized/nested/topology.rs]: Can we change this to return dx and dy? Right now this computes the numpy-like operations to transform data, but this makes the API targeted to a very specific use-case (and experience tells me that this would be very brittle). With dx and dy you should be able to infer all the necessary information to compute the transformation yourself.

That also means that we need to rename the function to something more appropriate, but I don't have any ideas right now.

Also, can we move this into `crate::scalar::nested`, please?

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+/// Split NESTED cell indexes into base-face-local coordinates.
+///
+/// Inputs are assumed to have already been validated. The returned vectors have
+/// the same length and contain `(face, x, y)` components in matching order.
```
Comment (keewis) [inline rust/healpix-geo-core/src/vectorized/nested/topology.rs]: I would follow cdshealpix' naming convention instead of healpy's:
```suggestion
/// the same length and contain `(face, i, j)` components in matching order.
```

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+use cdshealpix::compass_point::MainWind;
+use cdshealpix::nested::get;
+use cdshealpix::nested::zordercurve::get_zoc;
+
+use crate::maybe_parallelize;
+use crate::vectorized::depth::Depth;
+
+#[inline]
+fn pix2xyf_scalar(hash: u64, depth: u8) -> (u8, u32, u32) {
```
Comment (keewis) [inline rust/healpix-geo-core/src/vectorized/nested/topology.rs]: please move this to `crate::scalar::nested` and call it something along the lines of `healpix_to_base_cell_coordinates` (to follow cdshealpix terminology)

Additionally, `cdshealpix` definitely implements this function, so it might be worth asking about making that public instead of reimplementing ourselves (since we'd be waiting on `cdshealpix` this can be a separate PR).

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        flip_x,
+        flip_y,
+    })
+}
+
+/// Split NESTED cell indexes into base-face-local coordinates.
+///
+/// Inputs are assumed to have already been validated. The returned vectors have
+/// the same length and contain `(face, x, y)` components in matching order.
+pub fn pix2xyf(ipix: &[u64], depth: Depth, nthreads: usize) -> (Vec<u8>, Vec<u32>, Vec<u32>) {
```
Comment (keewis) [inline rust/healpix-geo-core/src/vectorized/nested/topology.rs]: to be consistent with the other API, this should return `Vec<(u8, u32, u32)>`. Also, this as well should be renamed to follow the existing naming convention (see my suggestion on the scalar function).

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        x.push(x_value);
+        y.push(y_value);
+    }
+    (face, x, y)
+}
+
+/// Combine base faces and face-local coordinates into NESTED cell indexes.
+///
+/// Inputs are assumed to have equal lengths and to have already been validated.
+pub fn xyf2pix(face: &[u8], x: &[u32], y: &[u32], depth: Depth, nthreads: usize) -> Vec<u64> {
```
Comment (keewis) [inline rust/healpix-geo-core/src/vectorized/nested/topology.rs]: this should take a `coords: &[(u8, u32, u32)]` to be consistent with the other API (see `crate::scalar::nested::coordinates::lonlat_to_healpix` for an example)

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs [hunk 9]
```diff
@@ -0,0 +1,217 @@
+//! Integer-only topology operations for NESTED cells.
```
Comment (keewis) [inline rust/healpix-geo-core/src/vectorized/nested/topology.rs]: Spelling the indexing scheme all-caps has always felt weird to me since "nested" is a proper english word and not an abbreviation.
```suggestion
//! Integer-only topology operations for nested cells.
```

**Review by keewis — DISMISSED** [review]
I have a few suggestions regarding the naming and the API, which has an effect on the implementation, too.

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs (hunk 4 above)
Comment (tinaok) [reply rust/healpix-geo-core/src/vectorized/nested/topology.rs]: yes that's fine for me, as long as we do not change the API on python side.

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs (hunk 6 above)
Comment (tinaok) [reply rust/healpix-geo-core/src/vectorized/nested/topology.rs]: That’s fine for me. I agree with moving this to crate::scalar::nested and renaming it on the Rust side to follow the existing terminology.
My main constraint is to keep the Python-facing API stable, since it is already exposed and used downstream. As long as the performance does not regress and the Python API remains unchanged, I’m also fine with a follow-up PR to reuse the cdshealpix implementation directly.

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs (hunk 8 above)
Comment (tinaok) [reply rust/healpix-geo-core/src/vectorized/nested/topology.rs]: I agree with adapting the Rust API to use coords: &[(u8, u32, u32)] for consistency with the rest of the Rust code.
For the Python bindings, I would prefer to keep the current NumPy-friendly API unchanged, because it is already called from healpix-analyse. So I’m fine with changing the Rust core API as long as we keep a compatibility layer on the Python side.

File: rust/healpix-geo-core/src/vectorized/nested/topology.rs (hunk 4 above)
Comment (keewis) [reply rust/healpix-geo-core/src/vectorized/nested/topology.rs]: while the comment was on the rust code, I do mean that we should change the python API as well.

The reason is that `dx` and `dy` are information computed from geometry, while the transform you're proposing is acting on that information. `healpix-geo` is a geometry library and does not interact with data at all so far, and this would be a significant departure from that boundary.

Since the results are scalars (mostly `bool`s) and the total number of combinations the parameters can have is small (`12 * 8 = 96`), I don't think it impacts performance either (and if it does this can easily be mitigated using `functools.lru_cache` or a different caching mechanism on the function that computes this).

**Comment by keewis** [comment]
We're still missing the same for `ring`, but I'm not sure this is applicable (we'd be splitting into ring number and in-ring position). Either way, this will need a lot more thought than I have time for right now.

the CI cache problems are unrelated (even though I tried to fix that here), so I'll go ahead and merge anyways.
